events/request.go

```diff
@@ -0,0 +1,165 @@
+	StartLast  Start = "LATEST"
+
+	// Start from a specific offset
+	StartOffset Start = ""
+)
+
+// DeviceType can be specified in a Filter to receive events for specific types
+// of devices.
+type DeviceType string
+
+const (
+	DeviceAmazon  DeviceType = "amazon"
+	DeviceAndroid DeviceType = "android"
+	DeviceIOS     DeviceType = "ios"
+	DeviceUser    DeviceType = "named_user"
+	deviceUnknown DeviceType = "unknown"
+)
+
+type Filter struct {
+	Types        []Type       `json:"type,omitempty"`
+	DeviceTypes  []DeviceType `json:"device_types,omitempty"`
+	Notification []Push       `json:"notification,omitempty"`
+	Devices      []Device     `json:"devices,omitempty"`
+	Latency      int64        `json:"latency,omitempty"`
+}
+
+type SubsetType string
+
+const (
+	SubsetTypePartition SubsetType = "PARTITION"
+	SubsetTypeSample    SubsetType = "SAMPLE"
+)
+
+type Subset struct {
+	Type       SubsetType `json:"type"`
+	Proportion *float64   `json:"proportion,omitempty"`
+	Count      *int       `json:"count,omitempty"`
+	Selection  *int       `json:"selection,omitempty"`
+}
+
+// SubsetPartition creates a Subset with "count" deterministic partitions of
+// which this request will receive the "selection" partition.
+func SubsetPartition(count, selection int) *Subset {
+	return &Subset{Type: SubsetTypePartition, Count: &count, Selection: &selection}
+}
+
+// SubsetSample creates a random sampling Subset whose proportion should be
+// between 0 and 1.
+func SubsetSample(proportion float64) *Subset {
+	return &Subset{Type: SubsetTypeSample, Proportion: &proportion}
+}
+
+// Validate returns an error if Subset is invalid otherwise nil.
+func (s *Subset) Validate() error {
+	if s == nil {
+		// It's valid to not specify a subset
+		return nil
+	}
+	switch s.Type {
+	case SubsetTypePartition:
+		if s.Count == nil || s.Selection == nil {
+			return errors.New("count and selection must be set for partition subsets")
+		}
+		if *s.Count < 1 {
+			return errors.New("count < 1")
+		}
+		if *s.Selection < 0 || *s.Selection >= *s.Count {
+			return fmt.Errorf("selection must be [0,%d)", *s.Count)
+		}
+		if s.Proportion != nil {
+			return errors.New("proportion must not be set for partition subsets")
+		}
+	case SubsetTypeSample:
+		if s.Proportion == nil {
+			return errors.New("proportion must be set for sample subsets")
+		}
+		if *s.Proportion < 0 || *s.Proportion > 1 {
+			return fmt.Errorf("proportion %f not between [0,1]", *s.Proportion)
+		}
+		if s.Count != nil || s.Selection != nil {
+			return errors.New("count and selection must not be set for sample subsets")
+		}
+	default:
+		return fmt.Errorf("invalid subset type: %s", s.Type)
+	}
+	return nil
+}
+
+// Request is an Urban Airship Events API request. The Fetch function will
+// create one internally, or you can manually create your own and submit it via
+// the gobyairship.Client's Post method.
+type Request struct {
+	// Start is one of “EARLIEST” or “LATEST”. Specifies that the stream should
```

Comment:
Honestly it's probably better to go with earliest/latest anyways because it's more explicit.